File: consensus/src/main/scala/co/topl/consensus/interpreters/ConsensusValidationState.scala
package co.topl.consensus.interpreters

import cats.data.OptionT
import cats.implicits._
import cats.{Applicative, MonadThrow}
import co.topl.algebras.ClockAlgebra.implicits._
import co.topl.algebras._
import co.topl.brambl.models.TransactionId
import co.topl.brambl.models.transaction.IoTransaction
import co.topl.consensus.algebras.ConsensusValidationStateAlgebra
import co.topl.consensus.models.{ActiveStaker, BlockId, StakingAddress}
import co.topl.eventtree.EventSourcedState
import co.topl.models._

object ConsensusValidationState {

  /**
   * Constructs a ConsensusValidationState using the combination of an EpochBoundary EventSourcedState and a ConsensusState
   * EventSourcecdState.  Requests for blocks at the tip of the chain will return data from 2 epochs before that block.
   */
  def make[F[_]: MonadThrow](
    genesisBlockId: BlockId,
    epochBoundaryEventSourcedState: EventSourcedState[F, EpochBoundariesEventSourcedState.EpochBoundaries[
      F
    ], BlockId],
    consensusDataEventSourcedState: EventSourcedState[F, ConsensusDataEventSourcedState.ConsensusData[
      F
    ], BlockId],
    clock:            ClockAlgebra[F],
    fetchTransaction: TransactionId => F[IoTransaction]
  ): F[ConsensusValidationStateAlgebra[F]] =
    Applicative[F].pure {
      new ConsensusValidationStateAlgebra[F] {
        def totalActiveStake(currentBlockId: BlockId, slot: Slot): F[BigInt] =
          useStateAtTargetBoundary(currentBlockId, slot)(_.totalActiveStake.getOrRaise(()))

        def staker(currentBlockId: BlockId, slot: Slot)(
          address: StakingAddress
        ): F[Option[ActiveStaker]] =
          OptionT(useStateAtTargetBoundary(currentBlockId, slot)(_.stakers.get(address)))
            .flatMapF(a => fetchTransaction(a.id).map(_.outputs.get(a.index)))
            .subflatMap(_.value.value.topl)
            .subflatMap(t => t.registration.map(ActiveStaker(_, t.quantity)))
            .value

        /**
         * Determines the N-2 epoch from the given block, then determines the final block ID of the N-2 epoch.  That
         * N-2 block is used in determining the `ConsensusState` to retrieve.  Once retrieved, it is applied to the
         * given `f` function
         */
        private def useStateAtTargetBoundary[Res](
          currentBlockId: BlockId,
          slot:           Slot
        )(f: ConsensusDataEventSourcedState.ConsensusData[F] => F[Res]): F[Res] =
          for {
            epoch <- clock.epochOf(slot)
            // Note: Blocks created within the first two epochs should use the state from the genesis block
            boundaryBlockId <-
              if (epoch > 1)
                epochBoundaryEventSourcedState.useStateAt(currentBlockId)(_.getOrRaise(epoch - 2))
              else
                genesisBlockId.pure[F]
            res <- consensusDataEventSourcedState.useStateAt(boundaryBlockId)(f)
          } yield res
      }
    }
}

File: dart/consensus/lib/interpreters/consensus_validation_state.dart
import 'package:bifrost_common/algebras/clock_algebra.dart';
import 'package:bifrost_common/algebras/event_sourced_state_algebra.dart';
import 'package:bifrost_common/models/common.dart';
import 'package:bifrost_consensus/algebras/consensus_validation_state_algebra.dart';
import 'package:bifrost_consensus/interpreters/consensus_data_event_sourced_state.dart';
import 'package:bifrost_consensus/interpreters/epoch_boundaries.dart';
import 'package:topl_protobuf/consensus/models/staking_address.pb.dart';
import 'package:topl_protobuf/consensus/models/operational_certificate.pb.dart';
import 'package:topl_protobuf/consensus/models/block_id.pb.dart';
import 'package:rational/rational.dart';
import 'package:fixnum/fixnum.dart';

class ConsensusValidationState extends ConsensusValidationStateAlgebra {
  final BlockId genesisBlockId;
  final EventSourcedStateAlgebra<EpochBoundariesState, BlockId>
      epochBoundaryState;
  final EventSourcedStateAlgebra<ConsensusData, BlockId> consensusDataState;
  final ClockAlgebra clock;

  ConsensusValidationState(this.genesisBlockId, this.epochBoundaryState,
      this.consensusDataState, this.clock);

  @override
  Future<SignatureKesProduct?> operatorRegistration(
          BlockId currentBlockId, Int64 slot, StakingAddress address) =>
      _useStateAtTargetBoundary(
          currentBlockId, slot, (t) => t.registrations.get(address));

  @override
  Future<Rational?> operatorRelativeStake(
          BlockId currentBlockId, Int64 slot, StakingAddress address) =>
      _useStateAtTargetBoundary(currentBlockId, slot, (consensusData) async {
        final maybeStake = await consensusData.operatorStakes.get(address);
        if (maybeStake != null) {
          final totalStake =
              await consensusData.totalActiveStake.getOrRaise({});
          return Rational(maybeStake, totalStake);
        }
      });

  _useStateAtTargetBoundary<Res>(BlockId currentBlockId, Slot slot,
      Future<Res> Function(ConsensusData) f) async {
    final epoch = clock.epochOfSlot(slot);
    final targetEpoch = epoch - 2;
    final boundaryBlockId = (targetEpoch >= 0)
        ? await epochBoundaryState.useStateAt(
            currentBlockId, (s) => s.getOrRaise(targetEpoch))
        : genesisBlockId;
    return consensusDataState.useStateAt(boundaryBlockId, f);
  }
}
